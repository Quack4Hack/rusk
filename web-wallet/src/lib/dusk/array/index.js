export { default as arraysEqual } from "./arraysEqual";
export { default as shuffleArray } from "./shuffleArray";
export { default as findFirstNMatches } from "./findFirstNMatches";
