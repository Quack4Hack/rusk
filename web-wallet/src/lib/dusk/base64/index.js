export { default as base64ToUTF8 } from "./base64ToUTF8";
export { default as base64ToBytes } from "./base64ToBytes";
export { default as bytesToBase64 } from "./bytesToBase64";
export { default as utf8ToBase64 } from "./utf8ToBase64";
