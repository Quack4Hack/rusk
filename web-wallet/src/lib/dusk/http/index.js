export { default as failureToRejection } from "./failureToRejection";
