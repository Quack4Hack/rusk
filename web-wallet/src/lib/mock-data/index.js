export { default as addresses } from "./addresses";
export { default as apiMarketData } from "./api-market-data.json";
export { default as stakeInfo } from "./stakeInfo";
export { default as transactions } from "./transactions";
