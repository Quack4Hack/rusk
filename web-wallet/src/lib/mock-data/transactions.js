/* eslint-disable camelcase */

/** @type {Transaction[]} */
export default [
  {
    amount: 25000,
    block_height: 33458,
    direction: "In",
    fee: 1876,
    id: "tg874534tddsghdtst54hdfh793h",
    tx_type: "TRANSFER",
  },
  {
    amount: -25000,
    block_height: 33456,
    direction: "In",
    fee: 1876,
    id: "tg874534tddsghdtst54hdfh791h",
    tx_type: "TRANSFER",
  },
  {
    amount: 25000,
    block_height: 33455,
    direction: "In",
    fee: 1876,
    id: "tg874534tddsghdtst54hdfh790h",
    tx_type: "STAKE",
  },
  {
    amount: -25000,
    block_height: 33459,
    direction: "In",
    fee: 1876,
    id: "tg874534tddsghdtst54hdfh794h",
    tx_type: "TRANSFER",
  },
  {
    amount: 27000,
    block_height: 33460,
    direction: "Out",
    fee: 1876,
    id: "tg874534tddsghdtst54hdfh795h",
    tx_type: "STAKE",
  },
  {
    amount: -25000,
    block_height: 33457,
    direction: "Out",
    fee: 1876,
    id: "tg874534tddsghdtst54hdfh792h",
    tx_type: "TRANSFER",
  },
];
