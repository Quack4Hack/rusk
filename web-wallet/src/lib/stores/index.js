export { default as operationsStore } from "./operationsStore";
export { default as settingsStore } from "./settingsStore";
export { default as walletStore } from "./walletStore";
