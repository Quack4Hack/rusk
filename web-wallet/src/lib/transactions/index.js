export { default as getLastTransactionHash } from "./getLastTransactionHash";
export { default as sortByHeightDesc } from "./sortByHeightDesc";
