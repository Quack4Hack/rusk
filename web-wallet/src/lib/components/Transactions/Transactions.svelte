<script>
  import { compose, take } from "lamb";
  import { mdiArrowLeft, mdiContain } from "@mdi/js";
  import { onMount } from "svelte";
  import { fade } from "svelte/transition";
  import { Badge, Icon, Suspense, Throbber } from "$lib/dusk/components";
  import {
    createFeeFormatter,
    createTransferFormatter,
  } from "$lib/dusk/currency";
  import { calculateAdaptiveCharCount, middleEllipsis } from "$lib/dusk/string";
  import { sortByHeightDesc } from "$lib/transactions";

  import AppAnchor from "../AppAnchor/AppAnchor.svelte";
  import AppAnchorButton from "../AppAnchorButton/AppAnchorButton.svelte";

  /** @type {String} */
  export let language;

  /** @type {Number | Undefined} */
  export let limit = undefined;

  const transferFormatter = createTransferFormatter(language);
  const feeFormatter = createFeeFormatter(language);

  /** @type {Promise<Transaction[]>} */
  export let items;

  /** @type {Boolean}*/
  export let isSyncing;

  /** @type {Error|null}*/
  export let syncError;

  /** @type {Number} */
  let screenWidth = window.innerWidth;

  /** @type {(transactions: Transaction[]) => Transaction[]} */
  const getOrderedTransactions = limit
    ? compose(take(limit), sortByHeightDesc)
    : sortByHeightDesc;

  onMount(() => {
    const resizeObserver = new ResizeObserver((entries) => {
      const entry = entries[0];

      screenWidth = entry.contentRect.width;
    });

    resizeObserver.observe(document.body);

    return () => resizeObserver.disconnect();
  });
</script>

<article in:fade|global class="transactions">
  <header class="transactions__header">
    <h3 class="h4">Transactions</h3>
    {#if limit}
      <AppAnchorButton
        className="transactions__footer-button"
        href="/dashboard/transactions"
        text="All transactions"
        variant="primary"
      />
    {:else}
      <AppAnchorButton
        className="transactions__footer-button"
        href="/dashboard"
        text="Back"
        variant="tertiary"
        icon={{ path: mdiArrowLeft }}
      />
    {/if}
  </header>
  <Suspense
    className="transactions-list__container"
    errorMessage="Error getting transactions"
    errorVariant="details"
    waitFor={items}
  >
    <svelte:fragment slot="pending-content">
      <div class="transactions-list__loading-container">
        {#if !isSyncing && !syncError}
          <Throbber />
        {:else}
          <p>Data will load after a successful sync.</p>
        {/if}
      </div>
    </svelte:fragment>
    <svelte:fragment slot="success-content" let:result={transactions}>
      {#if transactions.length}
        {#each getOrderedTransactions(transactions) as transaction (transaction.id)}
          <dl class="transactions-list">
            <dt class="transactions-list__term">Hash</dt>
            <dd class="transactions-list__datum transactions-list__datum--hash">
              <samp>
                <AppAnchor
                  href="https://explorer.dusk.network/transactions/transaction?id={transaction.id}"
                  rel="noopener noreferrer"
                  target="_blank"
                >
                  {middleEllipsis(
                    transaction.id,
                    calculateAdaptiveCharCount(screenWidth, 320, 640, 5, 20)
                  )}
                </AppAnchor>
              </samp>
            </dd>
            {#if transaction.tx_type}
              <dt class="transactions-list__term">Type</dt>
              <dd class="transactions-list__datum">
                <Badge
                  className="transactions-list__badge"
                  text={transaction.tx_type}
                />
              </dd>
            {/if}
            <dt class="transactions-list__term">Block</dt>
            <dd class="transactions-list__datum">
              {new Intl.NumberFormat(language).format(transaction.block_height)}
            </dd>
            <dt class="transactions-list__term">Amount</dt>
            <dd class="transactions-list__datum">
              {transferFormatter(transaction.amount)}
              <span class="transactions-list__ticker">Dusk</span>
            </dd>
            {#if transaction.direction === "Out"}
              <dt class="transactions-list__term">Fee</dt>
              <dd class="transactions-list__datum">
                {feeFormatter(transaction.fee)}
                <span class="transactions-list__ticker">Dusk</span>
              </dd>
            {/if}
          </dl>
        {/each}
      {:else}
        <div class="transactions-list__empty">
          <Icon path={mdiContain} size="large" />
          <p>You have no transaction history</p>
        </div>
      {/if}
    </svelte:fragment>
  </Suspense>
</article>

<style lang="postcss">
  .transactions {
    border-radius: 1.25em;
    background: var(--surface-color);
    display: flex;
    flex-direction: column;
    gap: var(--default-gap);
    padding-top: 1.375em;

    &__header {
      display: flex;
      flex-direction: row;
      align-items: center;
      justify-content: space-between;
      padding: 0 1rem;
      gap: 0.625rem;
      flex-wrap: wrap;

      & :global(h3) {
        line-height: 150%;
      }
    }
  }

  :global {
    .transactions-list {
      display: grid;
      grid-template-columns: max-content auto;
      width: 100%;

      &__term {
        background-color: var(--background-color-alt);
        grid-column: 1;
        line-height: 130%;
        text-transform: capitalize;
        padding: 0.3125em 0.625em 0.3125em 1.375em;
      }

      &__ticker {
        text-transform: uppercase;
      }

      &__loading-container {
        margin: 1.25em 0;
      }

      &__datum {
        grid-column: 2;
        line-height: 150%;
        padding: 0.312em 0.625em;
        display: flex;
        align-items: center;
        gap: 0.625em;
        font-family: var(--mono-font-family);
        overflow: hidden;

        & samp {
          display: block;
          white-space: nowrap;
          overflow: hidden;
        }

        &--hash {
          justify-content: center;
        }
      }

      &__empty {
        display: flex;
        flex-direction: column;
        align-items: center;
        gap: 0.5em;
        margin: 1.25em 0;
      }

      &__badge {
        flex: 1;
      }

      & dt:first-of-type,
      & dd:first-of-type {
        padding-top: 1em;
      }

      & dt:last-of-type,
      & dd:last-of-type {
        padding-bottom: 1em;
      }

      & dt:first-of-type {
        border-top-right-radius: 2em;
      }

      & dt:last-of-type {
        border-bottom-right-radius: 2em;
      }
    }
  }
</style>
