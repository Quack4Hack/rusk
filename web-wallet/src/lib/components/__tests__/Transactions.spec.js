import { cleanup, render } from "@testing-library/svelte";
import { afterAll, afterEach, describe, expect, it, vi } from "vitest";
import { get } from "svelte/store";
import { base } from "$app/paths";

import { resolveAfter } from "$lib/dusk/promise";
import { settingsStore } from "$lib/stores";
import { transactions } from "$lib/mock-data";
import { sortByHeightDesc } from "$lib/transactions";
import {
  createFeeFormatter,
  createTransferFormatter,
} from "$lib/dusk/currency";

import Transactions from "../Transactions/Transactions.svelte";

global.ResizeObserver = vi.fn().mockImplementation(() => ({
  disconnect: vi.fn(),
  observe: vi.fn(),
  unobserve: vi.fn(),
}));

vi.useFakeTimers();

describe("Transactions", () => {
  const transactionsPromise = resolveAfter(1000, transactions);
  const emptyTransactionsPromise = resolveAfter(1000, []);
  const blockExplorerBaseUrl = "/explorer/transactions/transaction?id=";
  const highestTransactionID = sortByHeightDesc(transactions)[0].id;
  const settings = get(settingsStore);
  const language = settings.language;
  const transferFormatter = createTransferFormatter(language);
  const feeFormatter = createFeeFormatter(language);

  const baseProps = {
    isSyncing: false,
    items: transactionsPromise,
    language,
    syncError: null,
  };

  afterEach(() => {
    cleanup();
  });

  afterAll(() => {
    vi.useRealTimers();
  });

  it("renders a loading indicator after a successful sync if the transaction promise isn't resolved yet", async () => {
    const props = {
      ...baseProps,
      isSyncing: true,
    };
    const { getByRole, getByText, rerender } = render(Transactions, {
      props: props,
    });
    const notice = getByText("Data will load after a successful sync.");

    expect(notice).toBeInTheDocument();

    await rerender({ ...baseProps });

    const spinner = getByRole("progressbar");

    expect(spinner).toBeInTheDocument();
  });

  it("renders transactions correctly when items are fulfilled", async () => {
    const props = {
      ...baseProps,
      limit: 1,
    };

    const { getByText, container } = render(Transactions, props);

    await vi.advanceTimersToNextTimerAsync();

    const transaction = sortByHeightDesc(transactions)[0];

    const transactionAmount = getByText(transferFormatter(transaction.amount));
    const transactionBlockHeight = getByText(
      new Intl.NumberFormat(language).format(transaction.block_height)
    );
    const transactionType = getByText(transaction.tx_type.toUpperCase());
    const transactionFee = getByText(feeFormatter(transaction.fee));

    expect(container.firstChild).toMatchSnapshot();

    expect(transactionAmount).toBeInTheDocument();
    expect(transactionBlockHeight).toBeInTheDocument();
    expect(transactionType).toBeInTheDocument();
    expect(transactionFee).toBeInTheDocument();
  });

  it("renders the correct amount of Transactions, as the limit supplied", async () => {
    const props = {
      ...baseProps,
      limit: 3,
    };

    const { getAllByText, container } = render(Transactions, props);

    await vi.advanceTimersToNextTimerAsync();

    expect(container.firstChild).toMatchSnapshot();

    const transactionHashes = getAllByText("Hash");

    expect(transactionHashes).toHaveLength(3);
  });

  it("renders the Transactions in descending order", async () => {
    const props = {
      ...baseProps,
    };
    const { container } = render(Transactions, props);

    await vi.advanceTimersToNextTimerAsync();

    const sortedTransactions = sortByHeightDesc(transactions);

    const transactionElements =
      container.querySelectorAll(".transactions-list");

    sortedTransactions.forEach((transaction, index) => {
      expect(transactionElements[index]).toHaveTextContent(transaction.id);
    });
  });

  it("renders the Transactions with the correct CTA to the Block Explorer", async () => {
    const props = {
      ...baseProps,
      limit: 1,
    };

    const { getByRole } = render(Transactions, props);

    await vi.advanceTimersToNextTimerAsync();

    const transactionHrefBlockExplorerAnchor = getByRole("link", {
      name: highestTransactionID,
    });

    expect(transactionHrefBlockExplorerAnchor).toBeInTheDocument();
    expect(transactionHrefBlockExplorerAnchor).toHaveAttribute(
      "href",
      `${blockExplorerBaseUrl}${highestTransactionID}`
    );
  });

  it("displays empty state when no transactions are present", async () => {
    const props = {
      ...baseProps,
      items: emptyTransactionsPromise,
    };

    const { getByText } = render(Transactions, props);

    await vi.advanceTimersToNextTimerAsync();

    const emptyState = getByText("You have no transaction history");

    expect(emptyState).toBeInTheDocument();
  });

  it('displays the "All transactions" CTA if limit is supplied', async () => {
    const props = {
      ...baseProps,
      items: emptyTransactionsPromise,
      limit: 1,
    };

    const { getByRole } = render(Transactions, props);
    const allTransactionAnchor = getByRole("link", {
      name: "All transactions",
    });

    expect(allTransactionAnchor).toBeInTheDocument();
    expect(allTransactionAnchor).toHaveAttribute(
      "href",
      `${base}/dashboard/transactions`
    );
  });

  it('displays the "Back" CTA if no limit is supplied', async () => {
    const props = {
      ...baseProps,
      items: emptyTransactionsPromise,
    };

    const { getByRole } = render(Transactions, props);
    const backAnchor = getByRole("link", { name: "Back" });

    expect(backAnchor).toBeInTheDocument();
    expect(backAnchor).toHaveAttribute("href", `${base}/dashboard`);
  });

  it("handles error state when items are rejected", async () => {
    const promiseReject = Promise.reject(new Error("An error has occurred"));
    const props = {
      ...baseProps,
      items: promiseReject,
    };

    const { getByText } = render(Transactions, props);

    await vi.advanceTimersToNextTimerAsync();

    const errorState = getByText("Error getting transactions");

    expect(errorState).toBeInTheDocument();
  });
});
