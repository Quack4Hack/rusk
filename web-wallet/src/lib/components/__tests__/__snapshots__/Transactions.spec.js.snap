// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Transactions > renders the correct amount of Transactions, as the limit supplied 1`] = `
<article
  class="transactions svelte-13w6rnl"
>
  <header
    class="transactions__header svelte-13w6rnl"
  >
    <h3
      class="h4"
    >
      Transactions
    </h3>
     
    <a
      aria-disabled="false"
      class="dusk-anchor dusk-anchor-button dusk-anchor-button--variant--primary dusk-anchor-button--size--default transactions__footer-button"
      href="/some-base-path/dashboard/transactions"
    >
      <span
        class="dusk-anchor-button__text"
      >
        All transactions
      </span>
      
    </a>
  </header>
   
  <div
    class="dusk-suspense transactions-list__container"
  >
    <dl
      class="transactions-list"
    >
      <dt
        class="transactions-list__term"
      >
        Hash
      </dt>
      <dd
        class="transactions-list__datum transactions-list__datum--hash"
      >
        <samp>
          <a
            class="dusk-anchor"
            href="https://explorer.dusk.network/transactions/transaction?id=tg874534tddsghdtst54hdfh795h"
            rel="noopener noreferrer"
            target="_blank"
          >
            tg874534tddsghdtst54hdfh795h
          </a>
        </samp>
         
      </dd>
      <dt
        class="transactions-list__term"
      >
        Type
      </dt>
      <dd
        class="transactions-list__datum"
      >
        <span
          class="dusk-badge dusk-badge--variant-neutral transactions-list__badge"
        >
          STAKE
        </span>
         
      </dd>
      <dt
        class="transactions-list__term"
      >
        Block
      </dt>
      <dd
        class="transactions-list__datum"
      >
        33,460
         
      </dd>
      <dt
        class="transactions-list__term"
      >
        Amount
      </dt>
      <dd
        class="transactions-list__datum"
      >
        +27,000.00
         
        <span
          class="transactions-list__ticker"
        >
          Dusk
        </span>
         
      </dd>
      <dt
        class="transactions-list__term"
      >
        Fee
      </dt>
      <dd
        class="transactions-list__datum"
      >
        1,876.00
         
        <span
          class="transactions-list__ticker"
        >
          Dusk
        </span>
         
      </dd>
    </dl>
    <dl
      class="transactions-list"
    >
      <dt
        class="transactions-list__term"
      >
        Hash
      </dt>
      <dd
        class="transactions-list__datum transactions-list__datum--hash"
      >
        <samp>
          <a
            class="dusk-anchor"
            href="https://explorer.dusk.network/transactions/transaction?id=tg874534tddsghdtst54hdfh794h"
            rel="noopener noreferrer"
            target="_blank"
          >
            tg874534tddsghdtst54hdfh794h
          </a>
        </samp>
         
      </dd>
      <dt
        class="transactions-list__term"
      >
        Type
      </dt>
      <dd
        class="transactions-list__datum"
      >
        <span
          class="dusk-badge dusk-badge--variant-neutral transactions-list__badge"
        >
          TRANSFER
        </span>
         
      </dd>
      <dt
        class="transactions-list__term"
      >
        Block
      </dt>
      <dd
        class="transactions-list__datum"
      >
        33,459
         
      </dd>
      <dt
        class="transactions-list__term"
      >
        Amount
      </dt>
      <dd
        class="transactions-list__datum"
      >
        -25,000.00
         
        <span
          class="transactions-list__ticker"
        >
          Dusk
        </span>
         
      </dd>
    </dl>
    <dl
      class="transactions-list"
    >
      <dt
        class="transactions-list__term"
      >
        Hash
      </dt>
      <dd
        class="transactions-list__datum transactions-list__datum--hash"
      >
        <samp>
          <a
            class="dusk-anchor"
            href="https://explorer.dusk.network/transactions/transaction?id=tg874534tddsghdtst54hdfh793h"
            rel="noopener noreferrer"
            target="_blank"
          >
            tg874534tddsghdtst54hdfh793h
          </a>
        </samp>
         
      </dd>
      <dt
        class="transactions-list__term"
      >
        Type
      </dt>
      <dd
        class="transactions-list__datum"
      >
        <span
          class="dusk-badge dusk-badge--variant-neutral transactions-list__badge"
        >
          TRANSFER
        </span>
         
      </dd>
      <dt
        class="transactions-list__term"
      >
        Block
      </dt>
      <dd
        class="transactions-list__datum"
      >
        33,458
         
      </dd>
      <dt
        class="transactions-list__term"
      >
        Amount
      </dt>
      <dd
        class="transactions-list__datum"
      >
        +25,000.00
         
        <span
          class="transactions-list__ticker"
        >
          Dusk
        </span>
         
      </dd>
    </dl>
    
    
  </div>
</article>
`;

exports[`Transactions > renders transactions correctly when items are fulfilled 1`] = `
<article
  class="transactions svelte-13w6rnl"
>
  <header
    class="transactions__header svelte-13w6rnl"
  >
    <h3
      class="h4"
    >
      Transactions
    </h3>
     
    <a
      aria-disabled="false"
      class="dusk-anchor dusk-anchor-button dusk-anchor-button--variant--primary dusk-anchor-button--size--default transactions__footer-button"
      href="/some-base-path/dashboard/transactions"
    >
      <span
        class="dusk-anchor-button__text"
      >
        All transactions
      </span>
      
    </a>
  </header>
   
  <div
    class="dusk-suspense transactions-list__container"
  >
    <dl
      class="transactions-list"
    >
      <dt
        class="transactions-list__term"
      >
        Hash
      </dt>
      <dd
        class="transactions-list__datum transactions-list__datum--hash"
      >
        <samp>
          <a
            class="dusk-anchor"
            href="https://explorer.dusk.network/transactions/transaction?id=tg874534tddsghdtst54hdfh795h"
            rel="noopener noreferrer"
            target="_blank"
          >
            tg874534tddsghdtst54hdfh795h
          </a>
        </samp>
         
      </dd>
      <dt
        class="transactions-list__term"
      >
        Type
      </dt>
      <dd
        class="transactions-list__datum"
      >
        <span
          class="dusk-badge dusk-badge--variant-neutral transactions-list__badge"
        >
          STAKE
        </span>
         
      </dd>
      <dt
        class="transactions-list__term"
      >
        Block
      </dt>
      <dd
        class="transactions-list__datum"
      >
        33,460
         
      </dd>
      <dt
        class="transactions-list__term"
      >
        Amount
      </dt>
      <dd
        class="transactions-list__datum"
      >
        +27,000.00
         
        <span
          class="transactions-list__ticker"
        >
          Dusk
        </span>
         
      </dd>
      <dt
        class="transactions-list__term"
      >
        Fee
      </dt>
      <dd
        class="transactions-list__datum"
      >
        1,876.00
         
        <span
          class="transactions-list__ticker"
        >
          Dusk
        </span>
         
      </dd>
    </dl>
    
    
  </div>
</article>
`;
